lib/internal/repl.js

```diff
@@ -192,31 +188,49 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
     timer = setTimeout(flushHistory, kDebounceHistoryMS);
   }
 
+  const historyLockFile = `${historyPath}.LCK`;
   function flushHistory() {
     timer = null;
     if (writing) {
       pending = true;
       return;
     }
-    writing = true;
-    const historyData = repl.history.join(os.EOL);
-    fs.write(repl._historyHandle, historyData, 0, 'utf8', onwritten);
-  }
-
-  function onwritten(err, data) {
-    writing = false;
-    if (pending) {
-      pending = false;
-      online();
-    } else {
-      repl._flushing = Boolean(timer);
-      if (!repl._flushing) {
-        repl.emit('flushHistory');
+    // Ref: https://github.com/nodejs/node/issues/1634
+    // opening a file for exclusive write should be atomic
+    // on non-networked POSIX systems.
+    fs.open(historyLockFile, 'wx', (err, fd) => {
+      if (err && err.code === 'EEXIST') {
+        pending = true;
+        repl.resume();
+        return;
       }
-    }
+      writing = true;
+      const lastLine = repl.history[0] + '\n';
+      fs.write(repl._historyHandle, lastLine, NaN, 'utf8', onwritten(fd));
+    });
   }
-}
 
+  function onwritten(fd) {
+    return (err, data) => {
+      writing = false;
+      if (pending) {
+        pending = false;
+        online();
+      } else {
+        repl._flushing = Boolean(timer);
+        if (!repl._flushing) {
+          repl.emit('flushHistory');
+        }
+      }
+      unlockHistoryFile(fd);
```

Comment:
And btw, this is how bash/readline does it:
- No locking whatsoever
- On exit, append the current session’s history to .bash_history
- After that, read the entire history file, truncate in memory to the desired number of lines, and then save again

This approach obviously racy but apparently that’s sufficiently irrelevant to just not care.